packages/agreement-process/src/services/readModelService.ts

```diff
@@ -318,51 +319,108 @@ async function searchTenantsByName(
 }
 
 // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
-function getDelegateAgreementsFilters(producerIds: TenantId[] | undefined) {
-  return producerIds && producerIds.length > 0
-    ? [
+const getDelegationLookup = (kind: DelegationKind) => [
+  {
+    $lookup: {
+      from: "delegations",
+      let: {
+        eserviceId: "$data.eserviceId",
+        consumerId: "$data.consumerId",
+        producerId: "$data.producerId",
+      },
+      pipeline: [
         {
-          $lookup: {
-            from: "delegations",
-            localField: "data.eserviceId",
-            foreignField: "data.eserviceId",
-            as: "delegations",
+          $match: {
+            $expr: {
+              $and: [
+                {
+                  $eq: ["$data.eserviceId", "$$eserviceId"],
+                },
+                {
+                  $eq: [
+                    "$data.delegatorId",
+                    kind === delegationKind.delegatedConsumer
+                      ? "$$consumerId"
+                      : "$$producerId",
+                  ],
+                },
+                {
+                  $eq: ["$data.kind", kind],
+                },
+                {
+                  $eq: ["$data.state", delegationState.active],
+                },
+              ],
+            },
           },
         },
-        {
-          $unwind: {
-            path: "$delegations",
-            preserveNullAndEmptyArrays: true,
-          },
+      ],
+      as: "delegations",
+    },
+  },
+];
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+function getProducerOrConsumerWithDelegateFilters(
+  producerIds: TenantId[],
+  consumerIds: TenantId[],
+  requesterId: TenantId
+) {
+  const requesterRole: "Consumer" | "Producer" | undefined =
+    producerIds.length === 1 && producerIds[0] === requesterId
+      ? "Producer"
+      : consumerIds.length === 1 && consumerIds[0] === requesterId
+      ? "Consumer"
+      : undefined;
+
+  return match([requesterRole, producerIds.length > 0, consumerIds.length > 0])
+    .with(["Producer", false, P.any], () => []) // Impossible case (producerIds must be set)
+    .with(["Consumer", P.any, false], () => []) // Impossible case (consumerIds must be set)
```

Comment:
Minor but maybe we can simplify this match because, given the check above, we already know for sure that if we have requesterRole "Producer" the producerIds array has a length exactly 1 (same for the "Consumer" case)